### DISCUSSION TITLE
React native:  UI lagging on submit

### DISCUSSION THREAD

**Discussion by amjadbouhouch** [react-hook-form/react-hook-form#4472] [Q&A] [answered]
Greetings!
I recently start using react-hook-form in my react-native's project. And I noticed a `heavy lag` even in `production`.
### Environments
`react-native: 0.63.4`
`react-hook-form: 6.15.4`
### Behavior
This behavior happens when I do some expensive js work inside `onSubmit` function `Api calls`, `setState`, or even `simple loop`.
The `animation` of the button stops when `onSubmit` function triggered and Js drops to -1 fps.

1. I've Two TextInputs (login and password).
`<Controller
            render={({ onChange, value }) => (
              <TextInput
                error={errors && errors.email}
                mode="outlined"
                label="Adresse e-mail"
                returnKeyType="next"
                onChangeText={onChange}
                value={value}
                autoCapitalize="none"
                autoCompleteType="email"
                textContentType="emailAddress"
                keyboardType="email-address"
              />
            )}
            control={control}
            name="email"
          />`
2.  Submit button.
`<Button
            loading={loading}
            disabled={loading}
            mode="contained"
            onPress={handleSubmit(onSubmit)}
          >
            Login
          </Button>`
3. onSubmit function:
`function onSubmit(data) {
    setState((prev) => ({ ...prev, loading: true }));
    authApi
      .signIn({
        email: data.email.trim(),
        password: data.password,
      })
      .then((res) => {
        const { accessToken } = res.data;
        logIn(accessToken);
      })
      .catch((error: ApiProblem) => {
        setState({ loading: false, error });
      });
  }`

![login](https://s4.gifyu.com/images/onSubmit-use-case.gif)

Thanks for all the amazing work!

**Comment by bluebill1049** [comment]
`Controller` should only re-render itself (isolated re-render), i would try to remove all those inputs and see if anything is improved and investigate what's causing the lag.

**Reply by amjadbouhouch** [reply]
Thank you for your answer!
I did exactly what you said, and I find out that `updating the state`  inside onSubmit function causes the lag and stops the animation for a second. I just update the state of the button with `isLoading` to disabled it when the submit triggered.
And I've tried the `7.0.0-rc.5` version and I faced the same problem.
I'll try to dig more into it.
Thank you again!

**Comment by amjadbouhouch** [comment, marked as answer]
I manage to resolve this issue by storing the data in a state like this:

1. onSubmit function
`const onSubmit = data =>
    setData(data);`
2. Move the logic inside useEffect hook:
`useEffect(() => {
    if (data.email === undefined || data.password === undefined) return;
    setState((prev) => ({ ...prev, loading: true }));
    authApi
      .signIn({
        email: data.email.trim(),
        password: data.password,
      })
      .then((res) => {
        const { accessToken } = res.data;
        logIn(accessToken);
      })
      .catch((error: ApiProblem) => {
        setState({ loading: false, error });
      });
  }, [data.email, data.password]);`
  
But I don't think this is the proper way.

**Reply by bluebill1049** [reply]
doesn't looks right to me. would be good to reproduce the problem in expo snack and find out the root cause.
